src/main/scala/zio/keeper/Cluster.scala

```diff
@@ -1,31 +1,267 @@
 package zio.keeper
 
-import java.net.InetAddress
+import java.math.BigInteger
+import java.util.UUID
 
-import zio.{ IO, ZIO }
+import zio._
+import zio.console.putStrLn
+import zio.keeper.Cluster.InternalProtocol.{ Iam, IamNewHere, PleaseIntroduceYourself }
+import zio.keeper.Error.{ HandshakeError, SendError }
+import zio.keeper.Error.NodeUnknown
+import zio.nio._
+import zio.nio.channels.{ AsynchronousServerSocketChannel, AsynchronousSocketChannel }
 import zio.stream.Stream
-import zio.keeper.Cluster._
 
-trait Cluster {
-  final def join: ZIO[Credentials with Discovery with Transport, Error, Unit] = ???
+import scala.reflect.ClassTag
+
+trait Cluster[A] {
+  def nodes: UIO[List[NodeId]]
+
+  def send[B <: A](data: B, receipt: NodeId): IO[Error, Unit]
+
+  def broadcast[B <: A](data: B)(implicit classTag: ClassTag[B]): IO[Error, Unit]
+
+  def receive: Stream[Error, Message[A]]
 }
 
 object Cluster {
 
+  sealed trait InternalProtocol
+
+  object InternalProtocol {
+
+    case object PleaseIntroduceYourself extends InternalProtocol
```

Comment:
`RequestIdentity`?

Reply:
names are hard :) I will change that to `RequestIdentity`